eth/protocols/snap/gentrie.go

```diff
@@ -0,0 +1,287 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package snap
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+// genTrie interface is used by the snap syncer to generate merkle tree nodes
+// based on a received batch of states.
+type genTrie interface {
+	// update inserts the state item into generator trie.
+	update(key, value []byte) error
+
+	// commit flushes the right boundary nodes if complete flag is true. This
+	// function must be called before flushing the associated database batch.
+	commit(complete bool) common.Hash
+}
+
+// pathTrie is a wrapper over the stackTrie, incorporating numerous additional
+// logics to handle the semi-completed trie and potential leftover dangling
+// nodes in the database. It is utilized for constructing the merkle tree nodes
+// in path mode during the snap sync process.
+type pathTrie struct {
+	owner common.Hash     // identifier of trie owner, empty for account trie
+	tr    *trie.StackTrie // underlying raw stack trie
+	first []byte          // the path of first committed node by stackTrie
+	last  []byte          // the path of last committed node by stackTrie
```

Comment:
Previously, the `last` was 

> The (hex-encoded without terminator) key of last inserted entry, tracked as right boundary.

Is this `last`, here, using the same semantics, or is there some difference?

Reply:
You use "last committed", not "last inserted". I find that  bit confusing, because the actual last element will never be committed by the stacktrie, unless an explicit commit is triggered. 
Hence, it's better to track the 'last inserted', since the 'last committed' will for the most part be just a regular node in the middle of the range

Reply:
I really don't get it. Won't every single value at some point be `last`, and thus the filtering will prevent every single parent-node from being written??

Reply:
Previously the `last` was the key of last **inserted** entry;
In this pr, `last` is the key of last **committed** entry, namely it's not on the right boundary;

Reply:
> I really don't get it. Won't every single value at some point be last, and thus the filtering will prevent every single parent-node from being written??

The right boundary filtering logic is changed a bit.

Previously, we invoke `Commit` function anyway regardless it's right complete or not, therefore the nodes on the path of "last inserted" node are filtered out;

In this pull request, we **won't** invoke `Commit` function if it's regarded as right-incomplete. The last inserted node along with its parent/ancestor on the right boundary are dropped without commit.